framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py

```diff
@@ -748,118 +789,171 @@ def set_monitor_equation(self, **data):
+        rendering_rules = SimulatorFragmentRenderingRules(next_form, SimulatorWizzardURLs.SET_PSE_RANGE_PARAMS_URL,
                                                           SimulatorWizzardURLs.SET_PSE_PARAMS_URL,
-                                                          last_form_url=SimulatorWizzardURLs.LAUNCH_PSE_URL)
+                                                          is_simulator_copy, is_simulator_load,
+                                                          last_form_url=self.last_loaded_form_url)
         return rendering_rules.to_dict()
 
     @cherrypy.expose
+    @using_template("simulator_fragment")
     @handle_error(redirect=False)
     @check_user
-    def launch_pse(self, **data):
-        # TODO: Split into: set range values and Launch, show message with finished config and nr of simulations
+    def set_pse_range_params(self, **data):
+        is_simulator_copy = common.get_from_session(common.KEY_IS_SIMULATOR_COPY) or False
+        is_simulator_load = common.get_from_session(common.KEY_IS_SIMULATOR_LOAD) or False
         all_range_parameters = self.range_parameters.get_all_range_parameters()
-        range_param1, range_param2 = SimulatorPSEParamRangeFragment.fill_from_post(all_range_parameters, **data)
+
+        if cherrypy.request.method == 'POST':
+            is_simulator_copy = False
+            self._update_last_loaded_fragment_url(SimulatorWizzardURLs.LAUNCH_PSE_URL)
+            range_param1, range_param2 = SimulatorPSEParamRangeFragment.fill_from_post(all_range_parameters, **data)
+            common.add2session(common.KEY_PSE_PARAM_1, range_param1)
+            common.add2session(common.KEY_PSE_PARAM_2, range_param2)
+        else:
+            range_param1, range_param2 = self._handle_range_params_at_loading()
+
+        number_of_simulations = len(range_param1.get_range_values())
+
+        if range_param2:
+            number_of_simulations = number_of_simulations * len(range_param2.get_range_values())
+
+        next_form = SimulatorPSEFinalFragment()
+        next_form.fill_from_post({'number_of_simulations': str(number_of_simulations)})
+
+        rendering_rules = SimulatorFragmentRenderingRules(next_form, SimulatorWizzardURLs.LAUNCH_PSE_URL,
+                                                          SimulatorWizzardURLs.SET_PSE_RANGE_PARAMS_URL,
+                                                          is_simulator_copy, is_simulator_load,
+                                                          last_form_url=self.last_loaded_form_url,
+                                                          is_launch_pse_fragment=True)
+        return rendering_rules.to_dict()
+
+    @cherrypy.expose
+    @handle_error(redirect=False)
+    @check_user
+    def launch_pse(self, **data):
         session_stored_simulator = common.get_from_session(common.KEY_SIMULATOR_CONFIG)
+        session_stored_simulator.simulation_length = float(common.get_from_session(common.KEY_SIMULATION_LENGTH))
```

Comment:
why keep length as a separate field in session? 
It should already be set on session_stored_simulator. Or why not?

Reply:
simulation_length is actually set in launch_simulation method. However, if you wish to run a PSE simulation, you don't go through launch_simulation. I actually had an idea that simulation name and length should be stored in a separate fragment from the last one with launch simulation and then in launch_simulation we just launch it, we don't send any parameter. That change would solve this problem too. Should I do this?

Reply:
Can't we stay with the same form (containing name and length) regardless if it is PSE or single op?

Reply:
In the way that it is now, launch_simulation gets name and simulation_length as parameters, launch_pse doesn't. If I make them both get these parameters (i.e. using the same form), then in case of launch_pse the form will be included twice in the flow, which we don't want. Something that I can think of is to make the user choose between normal and pse simulation BEFORE that form appears.